### PR TITLE
Add direct filesystem acces mode to dune cache.

### PR DESCRIPTION
**Pull request by mefyl**
This is a prerequisite for #2754 .

### CODE REVIEW

File: src/dune_memory/dune_memory_intf.ml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  { directory : string
+  ; remote : string
+  ; commit : string
+  }
+
+type command = Dedup of (Path.Build.t * Path.t * Digest.t)
+
+type handler = command -> unit
+
+module type memory = sig
```
Comment (rgrinberg) [inline src/dune_memory/dune_memory_intf.ml]: Should be `Memory`.

**Comment by mefyl** [comment]
Addressed @rgrinberg , thanks for the review !

File: src/dune/build_system.ml [hunk 2]
```diff
@@ -409,10 +409,16 @@ module Context_or_install = struct
     | Context s -> Context_name.to_dyn s
 end
 
+module type caching = sig
```
Comment (rgrinberg) [inline src/dune/build_system.ml]: No module types that start with a lowercase letter please :)

File: src/dune_memory/dune_memory.ml [hunk 3]
```diff
@@ -22,14 +15,23 @@ let key_of_string s =
...
     Printf.sprintf "%s promoted as %s"
-      (Path.Local.to_string (Path.Build.local original))
-      (Path.to_string promoted)
+      (Path.Local.to_string (Path.Build.local in_the_build_directory))
+      (Path.to_string in_the_memory)
+
+let command_to_dyn = function
+  | Dedup { in_the_build_directory; in_the_memory; digest } ->
+    let open Dyn.Encoder in
+    constr "Dedup"
```
Comment (rgrinberg) [inline src/dune_memory/dune_memory.ml]: small nit, but I believe `record` is a better argument here.

File: src/dune_memory/dune_memory.ml [hunk 4]
```diff
@@ -141,10 +120,8 @@ module Memory = struct
   let path_tmp memory = Path.L.relative memory.root [ "temp" ]
 
   let with_lock memory f =
-    let lock =
-      Stdune.Lock_file.create (Path.L.relative memory.root [ ".lock" ])
-    in
-    let finally () = Stdune.Lock_file.unlock lock in
+    let lock = Lock_file.create (Path.L.relative memory.root [ ".lock" ]) in
```
Comment (rgrinberg) [inline src/dune_memory/dune_memory.ml]: `Path.relative` is simpler here

File: bin/import.ml [hunk 5]
```diff
@@ -28,14 +28,41 @@ module Cached_digest = Dune.Cached_digest
...
+  | Daemon
+  | Direct
+
+let get_cache_mode () =
+  let var = "DUNE_CACHE_MODE" in
+  match Env.get Env.initial var with
+  | None
+  | Some "DAEMON" ->
+    Daemon
+  | Some "DIRECT" -> Direct
```
[the remark this answers is not in the document]
Comment (mefyl) [reply bin/import.ml]: Surely, you know better about the interface. I've made them case insensitive, if that's fine.

**Comment by rgrinberg** [comment]
Let's just allow lowercase for now. We don't allow case insensitivity anywhere else in dune. So I feel like it's a bit inconsistent to just allow it in this one instance.

**Pull request merged**